File: go/recursive/permutations_ii.go
package main

import "sort"

//全排列II

func permuteUnique(nums []int) [][]int {
	sort.Ints(nums)
	var res [][]int
	var data []int
	visited := make([]bool, len(nums))
	permuteUniqueHelper(nums, visited, data, &res)
	return res
}

func permuteUniqueHelper(nums []int, visited []bool, data []int, res *[][]int) {
	if len(data) == len(nums) {
		temp := make([]int, len(nums))
		copy(temp, data)
		*res = append(*res, temp)
		return
	}

	for i, v := range nums {
		if (i > 0 && nums[i] == nums[i-1] && !visited[i-1]) || visited[i] {
			continue
		}
		data = append(data, v)
		visited[i] = true
		permuteUniqueHelper(nums, visited, data, res)
		visited[i] = false
		data = data[:len(data)-1]
	}
}

File: python/backtracking/permutations_ii.py
from typing import List


# 全排列II

class Solution:

    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res, visited = [], [False] * len(nums)
        self.helper(nums, visited, [], res)
        return res

    
    def helper(self, nums, visited, data, res):
        if len(data) == len(nums):
            res.append(data[:])
            return

        for i in range(len(nums)):
            if (i > 0 and nums[i] == nums[i - 1] and not visited[i - 1]) or visited[i]:
                continue
            visited[i] = True
            data.append(nums[i])
            self.helper(nums, visited, data, res)
            data.pop()
            visited[i] = False


if __name__ == "__main__":
    print(1)